caffe2/contrib/tensorrt/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,224 @@
+  }
+  for (int i = 0; i < operator_def.output_size(); ++i) {
+    outputs.emplace(operator_def.output(i), i);
+    VLOG(0) << "Adding Output: " << operator_def.output(i);
+  }
+
+  // Set up the output size hints
+  std::vector<int> output_size_hints_encoded(
+      OperatorBase::GetRepeatedArgument<int>("output_size_hints"));
+  std::vector<std::string> output_size_names(
+      OperatorBase::GetRepeatedArgument<std::string>("output_size_names"));
+  int idx = 0;
+  for (const auto& oname : output_size_names) {
+    const auto it = outputs.find(oname);
+    if (it != outputs.end()) {
+      std::vector<TIndex> dims;
+      for (; idx < output_size_hints_encoded.size() && output_size_hints_encoded[idx] > 0; ++idx) {
+        dims.push_back(output_size_hints_encoded[idx]);
+      }
+      output_size_hints_.emplace(it->second, std::move(dims));
+    }
+  }
+
+  // match and bind the input/output
+  int num_bindings = trt_engine_->getNbBindings();
+  for (int b = 0; b < num_bindings; ++b) {
+    const auto& name = trt_engine_->getBindingName(b);
+    nv_dims_.push_back(trt_engine_->getBindingDimensions(b));
+    if (trt_engine_->bindingIsInput(b)) {
+      const auto it = inputs.find(name);
+      CAFFE_ENFORCE(it != inputs.end(), MakeString("Cannot find trt input: ", name));
+      binding_hints_.emplace_back(it->second, true);
+    } else {
+      const auto it = outputs.find(name);
+      CAFFE_ENFORCE(it != outputs.end());
+      binding_hints_.emplace_back(it->second, false);
+    }
+  }
+
+  trt_executor_ = InferObject(trt_engine_->createExecutionContext());
+}
+
+void TensorRTOp::MaybeAdjustOutputShape(int output_idx, std::vector<TIndex>* dims) {
+  const auto it = output_size_hints_.find(output_idx);
+  if (it != output_size_hints_.end()) {
+    const auto& dims_hint = it->second;
+    auto total_trt = std::accumulate(dims->begin(), dims->end(), (TIndex)(1), std::multiplies<TIndex>());
+    auto total_c2 = std::accumulate(dims_hint.begin(), dims_hint.end(), (TIndex)(1), std::multiplies<TIndex>());
+    if (total_c2 != total_trt) {
+      LOG(WARNING) << "The total size of TensorRT op output and hint don't match: " << total_trt << " vs " << total_c2;
+      return;
+    }
+
+    bool identical_shape = true;
+    if (dims->size() != dims_hint.size()) {
+      identical_shape = false;
+    } else {
+      for (int i = 0; i < dims->size(); ++i) {
+        if((*dims)[i] != dims_hint[i]) {
+          identical_shape = false;
+          break;
+        }
+      }
+    }
+
+    // We conform to the output shape hints. NB: We might need an explicit reshape op for this
+    if (!identical_shape) {
+      *dims = dims_hint;
+    }
+  }
+}
+
+bool TensorRTOp::RunOnDevice() {
+  CAFFE_ENFORCE(trt_executor_);
+  // Decide input batch size
+  size_t N = 0;
+  bool first = true;
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    if (first) {
```

Comment:
You could use `if (i == 0)` and remove `first` variable.

Reply:
BTW, do you need to check that first dimension (batch size) does not exceed `batch_size_max` passed in OperatorDef? Or would TensorRT resize automatically?